### DISCUSSION TITLE
Does Bitbucket Cloud Support Project Filters?

### DISCUSSION THREAD

**Discussion by jjathman** [renovatebot/renovate#16901] [Request Help]
Our company is migrating from Bitbucket server to Bitbucket cloud. One thing we use heavily right now are "projects" within Bitbucket. With the server version of Bitbucket we setup auto discover filters like "projectName/*" which allows us to run renovate against a handful of repositories at a time.

It looks like that isn't how Bitbucket Cloud works as the repository names that the auto discover filter works on take the form "workspaceName/repoName". I do see the project information returned in the API call that Renovate makes, but I don't believe anything looks at this data.

Is project filtering supported with Bitbucket Cloud?

**Comment by PhilipAbed** [comment]
autodiscover should work on all platforms,

i searched for other discussions like this for you, 
hope this helps:
https://github.com/renovatebot/renovate/discussions/16111

confirmation for your question: https://github.com/renovatebot/renovate/discussions/14779

**Reply by jjathman** [reply]
Thank you for the links, those are similar issues but they don't quite answer my question.

In Bitbucket Server every repo is located within a project. The repository names that Renovate receives in the API call take the form "projectKey/repoName". However, in Bitbucket Cloud the structure is a little different. The top level is called a "workspace" which then has projects under it, and finally repositories. However, the API call that Renovate makes gets repository names that only include the workspace and repo name - "workspace/repoName". However, I would still like to filter based on the project key as we had been doing with Bitbucket Server (on-prem version). 

Hope that helps explain the issue we are seeing.

**Reply by PhilipAbed** [reply]
i think renovate filters by the repo name which is `workspace/repoName`, not the project name

**Reply by PhilipAbed** [reply]
you mean if you had something like this
![image](https://user-images.githubusercontent.com/42116482/182402590-45b03f35-5ae6-4f74-a02d-8dfd7d5579fe.png)

you want to filter by the name above, `Philip Abed` but renovate doesn't even see that

**Reply by jjathman** [reply]
Correct, in my example the name that Renovate sees would be `surescripts/sample-lib` but I want to do filtering by the project instead. 

<img width="294" alt="Screen Shot 2022-08-02 at 9 51 00 AM" src="https://user-images.githubusercontent.com/6406/182404880-b5761d7a-6a74-4a2a-9afc-d7a356a76c16.png">


When Renovate runs it calls the `https://api.bitbucket.org/2.0/repositories?role=contributor&pagelen=100`. This returns a data structure like this:

<img width="356" alt="image" src="https://user-images.githubusercontent.com/6406/182406055-a46f769b-9e8b-4ac1-b3bd-c2045c3c13f3.png">

I collapsed parts which are not important, but you can see that the project information is available in the returned API response so I think it would be possible to add a project filter.

**Reply by viceice** [reply]
Maybe we should add the project part, so we get `<workspace>/<project>/<repo>` in future. Would be a breaking change.

But if i remember you can't have `w1/p1/r1` and `w1/p2/r1`. Repo names need to be usinque inside workspace

**Reply by jjathman** [reply]
@viceice yes I think you are correct, the repository names need to be unique within a workspace. 

I did find one alternative to client side filtering which is a query parameter can be added to the auto discovery API call to do the filtering on the server. The example would be:

`https://api.bitbucket.org/2.0/repositories?role=contributor&pagelen=100&q=project.key="CPI"`
instead of
`https://api.bitbucket.org/2.0/repositories?role=contributor&pagelen=100`

**Reply by viceice** [reply]
it's not so easy as you think, because the current filter doesn't know any platform specific structures. it simply gets the discovered repos as strings.

so any change will probably break existing users.

**Reply by jjathman** [reply]
Agreed, I think changing the repositories returned to include the project key would be a more invasive change. Would being able to provide a custom autoDiscovery path be a viable option instead? Right now the path is hardcoded [here](https://github.com/renovatebot/renovate/blob/e4dbd4ad491f656aaa7f9da9d4dae2c18622a1e6/lib/modules/platform/bitbucket/index.ts#L102), but if those was configurable I could append on the query parameter I outlined above. I think nothing else would need to change then? The subsequent processing would continue to work as expected and if this parameter is optional I don't think any existing users would notice.

**Reply by rarkins** [reply]
Aren't projects independent if workspaces? Eg repos from different workspaces can belong to  the same project?

If so then I'd prefer to make this a first class filter option instead of changing to a structure of workspace/project/repo

**Reply by jjathman** [reply]
I don't think that's true. To create a project you must create it within a workspace (at least according to their API). https://developer.atlassian.com/cloud/bitbucket/rest/api-group-projects/#api-workspaces-workspace-projects-post

**Reply by rarkins** [reply]
Can repos belong to more than one project?

**Reply by jjathman** [reply]
I don't believe so, when adding a repository to a workspace the project is a dropdown which only allows a single selection. Also, in the API response from the endpoint that Renovate calls, the "project" element is not an array, it is just a single object.

**Reply by setchy** [reply]
This could be a useful feature for y organizations workflow.

We have a handful of projects that are used to group repositories that don't require maintenance (archived, pocs, temporary for testing etc)
